Fix #5562 Improve HTTP Field cache allocation

gregw opened the pull request:
Fix #5562 by initially putting cacheable fields into a inexpensive arraylist.
Only create the Trie (with space and complexity costs) if a second request is received.

Edit: the original PR only created the Trie on the second request, but that behaviour was removed in reviews. The cache is now created every request, even for non persistent requests.

gregw committed:
Fix #5562 Improve HTTP Field cache allocation

Fix #5562 by initially putting cacheable fields into a inexpensive arraylist.
Only create the Trie (with space and complexity costs) if a second request is received.

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5562-improve-field-cache

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5562-improve-field-cache

gregw committed:
Fixed NPE

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -231,7 +235,7 @@
         for (HttpHeader h : HttpHeader.values())
         {
-            if (!CACHE.put(new HttpField(h, (String)null)))
+            if (!h.asString().startsWith(":") && !CACHE.put(new HttpField(h, (String)null)))
```
I would rather have `HttpHeader.isPseudo()` instead.

gregw replied:
OK

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -884,10 +888,8 @@ else if (n == HttpTokens.LINE_FEED)
-                            if (_fieldCache == null && _version.getVersion() >= HttpVersion.HTTP_1_1.getVersion() && headerCache > 0)
-                                _fieldCache = new ArrayTernaryTrie<>(headerCache);
+                            // Should we cache header fields?
+                            getFieldCache();
```
Ehm, this is not readable. It's the call to a getter method, whose return value is not used, but the comment seem to imply an `if` statement?
The method should renamed to something that is more intuitive -- now I have to look at the implementation to understand why there is this naked getter call without using the return value.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -1052,18 +1058,42 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
+                                _cacheableFields = new ArrayList<>();
+                            if (_field == null)
+                                _field = new HttpField(_header, caseInsensitiveHeader(_headerString, _header.asString()), _valueString);
+                            _cacheableFields.add(_field);
```
In this branch we have a cache size > 0, we are HTTP/1.1 and there is no `Connection: close`, so chances are that this parser will see another request, likely more than 1.
I would remove entirely the `_cacheableFields` handling, and just use the trie -- we have already optimized its creation.

gregw replied:
The problem is that we don't know if we will get a `Connection: close` header before or after any cacheable fields.... which is why I want the list.
If Connection:close is the last field, we could have allocated a big cache, done the work to fill it up and then find out at the last moment that it is not persistent.

sbordet replied:
A client sending HTTP/1.1 first requests with `Connection: close` is not a common case, so you are right that we may see the `Connection: close` header after we have created the trie, but that would be a rare case for which I don't think the extra complexity is warranted.

gregw replied:
I don't know why you think it is a rare case?   For clients sending a 1 time HTTP/1.1 request we have no data on if the cacheable fields like`User-Agent` and `Cookie` will come before or after a `Connection: close`.

Creating the list means that we have minimal cost for all 1 request connections and only slight extra cost for the first request of a multiple request connections.
Not creating the list means that we have significant cost for an unknown proportion of 1 request connections and slight saving on multiple request connections.

So while I don't like the list that much, I think it is best average solution.

An alternative would be to always make the list, then when a new request comes in we can look at what we have cached and what is in the list.  We can then really see if fields are likely to be the same or if they are actually changing.  We could then make smarter decisions about full caches and this perhaps have smaller caches.   I have to think about this a bit more, but having the list on the first request is a step towards this if it is workable.

gregw replied:
If we always created the list of cacheable fields and then only created/updated the actual cache on the next request then we would have a bit more info.
If the cache is perfect (ie contains all the cacheable fields that exist and they never change between requests), then on subsequent requests the cacheableField list will be empty.
On the 2nd and later requests if there are fields in the list then there are several possibilities
 + If the field name is not in the cache, then either the cache is full or it is a new field.   In the later case we can add the field.
 + If the field name is in the cache, then it is a changed value for that field.  Either that field is going to continually change (eg digest auth), it has changed once and will be the same afterwards, it is one of several values that may alternate (eg User-Agent on an aggregated connection).   

Hmmm not sure there is anything smart we can do with this information.  I think it is just add the filed to the cache if there is space.   So always having the list might save a little bit of work on the Nth request of a N request stream.  Not worth it.

So I still think the list of the 1st request a reasonable compromise.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -1052,18 +1058,42 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
+                    {
+                        if (_handler.getHeaderCacheSize() <= 0 || (_version != null && _version.getVersion() != HttpVersion.HTTP_1_1.getVersion()))
+                            // Don't cache any fields
+                            _cacheableFields = NO_CACHE;
```
Add braces to this `if` block.

sbordet reviewed (changes requested):
I'm not particularly fond of this change.

I am positive on delaying the creation of the Trie, until we know it's a HTTP/1.1 and we do have cache size > 0.

Seems to me that the chances for caching are the most common case by far, so I would skip the temporary list and go straight to the trie.

gregw committed:
Feedback from review

Create `HttpHeader.isPseudo()`` method
improved clarity with `createFieldCacheIfNeeded()``

sbordet commented:
@gregw also, I think we need some test -- something that verifies that the field is `==` in subsequent requests, etc.

gregw commented:
@sbordet There are cache tests and they were broken until I fixed `getFieldCache()` to create the cache if needed.  I think that method only really exists for the tests.

gregw committed:
Feedback from review

Only defer Trie creation to first cacheable field, not until next request.

gregw commented:
@lorban see the emptyTrie method I've added here.   Also I wonder why the HttpField cache uses the Terniary Trie rather than just the array Trie.

I think this is one of the most important (if not THE important) case to optimise Tries for.  We basically have the static and dynamic Tries... it is almost like HTTP/2 Hpack.
I wonder if we'd be better off with a specific cache class rather than generic Tries ?

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -1034,14 +1031,14 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
                         // Don't cache headers if not persistent
-                        if (HttpHeaderValue.CLOSE.is(_valueString) || new QuotedCSV(_valueString).getValues().stream().anyMatch(HttpHeaderValue.CLOSE::is))
-                            _fieldCache = null;
+                        if (_handler.getHeaderCacheSize() > 0 && (HttpHeaderValue.CLOSE.is(_valueString) || _valueString.contains(HttpHeaderValue.CLOSE.asString())))
```
`String.contains()` is case sensitive, so this test would fail for `_valueString="Close,Other"`, right?
I think using the previous version was expensive, but it was on the last expression of the `||` so executed only rarely.

gregw replied:
Let me convert to the HttpField.contains method, which is not expensive

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/AbstractTrie.java:
```diff
@@ -81,4 +82,57 @@ public boolean isCaseInsensitive()
+
+    public static <T> Trie<T> emptyTrie(boolean caseInsensitive)
+    {
+        return new AbstractTrie<T>(caseInsensitive)
```
If it's empty, should be irrelevant if it's case sensitive, so perhaps we can make this anonymous class a `static final`?
Method `isCaseSensitive()` is only used internally by the implementation (e.g. by `put(...)`, and since this is the empty implementation it should not matter.
Can it be moved to the interface, something like `Trie.EMPTY`?

gregw replied:
A constant like `Trie.EMPTY` does not work because it is a generic type, so you need a method.  Thus may as well have the caseinsensitive field.   But I've moved to the interface.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -1052,18 +1049,29 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
-                    _fieldCache.put(_field);
+                    if (_fieldCache == null)
+                    {
+                        _fieldCache = (_handler.getHeaderCacheSize() > 0 && (_version != null && _version.getVersion() == HttpVersion.HTTP_1_1.getVersion()))
```
Can't the "version" expression be simplified to just: `(_version == HttpVersion.HTTP_1_1)`?

gregw committed:
Updates from review

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Trie.java:
```diff
@@ -131,4 +131,99 @@
+            @Override
+            public Set<String> keySet()
+            {
+                return null;
```
I would return an empty `Set` here.

lorban commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpHeader.java:
```diff
@@ -188,6 +195,11 @@ public boolean is(String s)
         return _string.equalsIgnoreCase(s);
     }
 
+    public boolean isPseudo()
```
This method requires a bit of javadoc to explain what a pseudo header is.

gregw replied:
done

lorban commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -1052,18 +1050,29 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
+                    if (_fieldCache == null)
+                    {
+                        _fieldCache = (_handler.getHeaderCacheSize() > 0 && (_version != null && _version == HttpVersion.HTTP_1_1))
+                            ? new ArrayTernaryTrie<>(_handler.getHeaderCacheSize())
```
The `getHeaderCacheSize` method name is a bit misleading as the `ArrayTernaryTrie` is going to consume a lot more bytes in heap than the specified size, which doesn't match the `HttpConfiguration`'s managed attribute doc that reads _"The maximum allowed size in bytes for an HTTP header field cache"_.

gregw replied:
updated javadoc

gregw committed:
Merge branch 'jetty-9.4.x' into jetty-9.4.x-5562-improve-field-cache

gregw committed:
Update from review

 + more javadoc
 + empty set return

gregw commented:
@lorban nudge
